pkg/util/ec2/ec2.go

```diff
@@ -169,31 +177,45 @@ func doHTTPRequest(url string, method string, headers map[string]string, retriab
 	if err != nil {
 		return nil, err
 	}
+
+	if useToken {
+		token, err := getToken()
+		if err != nil {
+			log.Info("ec2_prefer_imdsv2 is set to true in configuration but the agent was unable to get a token: %s", err)
+		} else {
+			headers["X-aws-ec2-metadata-token"] = token
+		}
+	}
+
 	for header, value := range headers {
 		req.Header.Add(header, value)
 	}
 
 	res, err := client.Do(req)
 	if err != nil {
 		return nil, err
-	}
-	if res.StatusCode == 401 && retriableWithFreshToken {
-		// Most of 401 errors can be solved by retrying with a fresh token
-		token, err := getToken()
-		if err != nil {
-			return nil, err
-		}
-		headers["X-aws-ec2-metadata-token"] = token
-		return doHTTPRequest(url, method, headers, false)
-
 	} else if res.StatusCode != 200 {
 		return nil, fmt.Errorf("status code %d trying to fetch %s", res.StatusCode, url)
 	}
-
 	return res, nil
 }
 
 func getToken() (string, error) {
+	token.RLock()
```

Comment:
Is a mutex needed? Can the call to the EC2 metadata endpoint occur somewhere else / parallel?

Reply:
The `ec2` package is used from several different places (for hostname resolution, clustername resolution, some docker logic, etc...) this package will never be heavily requested, but I think concurrent calls may happen, and I think nothing should prevent that.